game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProMoveUtils.java

```diff
@@ -317,11 +370,46 @@ public static void calculateAmphibRoutes(
 
         // Unload transport
         if (!loadedUnits.isEmpty() && !t.isWater()) {
-          moveUnits.add(loadedUnits);
-          transportsToLoad.add(null);
           final Route route = new Route(transportTerritory, t);
-          moveRoutes.add(route);
+          moves.add(new Move(loadedUnits, route));
+        }
+      }
+    }
+
+    // Re-order and batch the moves. That is, move all the units together, then
+    // move all the transports together, then unload them all.
+    for (final ArrayList<Move> moves : movesMap.values()) {
+      // First, add all the transport loads.
+      int i = 0;
+      for (final Move move : moves) {
+        if (move != null && move.isTransportLoad()) {
+          // Find all others with the same route to merge with.
+          mergeMoves(move, moves, i + 1);
+          move.addTo(moveUnits, moveRoutes, transportsToLoad);
+          moves.set(i, null);
         }
+        i++;
+      }
+
+      // Then, add all the transport and unload moves, merging moves together.
+      // Since we process the moves in order, no special logic is needed to make
+      // sure transports move before unloading.
+      i = 0;
+      for (final Move move : moves) {
+        if (move != null) {
+          mergeMoves(move, moves, i + 1);
+          move.addTo(moveUnits, moveRoutes, transportsToLoad);
+        }
+        i++;
+      }
+    }
+  }
+
+  private static void mergeMoves(
+      final Move move, final ArrayList<Move> moves, final int startIndex) {
+    for (int i = startIndex; i < moves.size(); i++) {
+      if (move.mergeWith(moves.get(i))) {
+        moves.set(i, null);
```

Comment:
Could you explain why you're explicitly nulling out the entries instead of clearing/removing them?

Reply:
Removing would make this loop O(n^2) instead of O(n).

Reply:
Fair enough, I missed the if case that only "removes" certain entries.

You might want to document that somewhere so future developers will be aware of that

Reply:
Done.